framework/src/bundle/BundleArchive.cpp

```diff
@@ -136,9 +141,27 @@ void BundleArchive::SetAutostartSetting(int32_t setting)
   data->autostartSetting = setting;
 }
 
-std::shared_ptr<BundleResourceContainer>
-BundleArchive::GetResourceContainer() const
+std::shared_ptr<BundleResourceContainer> BundleArchive::GetResourceContainer()
+  const
 {
   return resourceContainer;
 }
+
+void BundleArchive::RegisterOpenedResource() const 
+{ 
+  numOpenResources++; 
+}
+
+void BundleArchive::UnregisterOpenedResource() const 
+{ 
+  numOpenResources--;
+  if (numOpenResources == 0) {
```

Comment:
there is a race between lines 157 and 158 where one thread can decrement ```numOpenResources``` and before it executes line 158 another thread executes line 157. Even within line 158 there is a race between the time it takes to retrieve the value in ```numOpenResources``` and compare it to ```0```. The race can be solved by using https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange or by adding a mutex around the increment, decrement and comparison of ```numOpenResources``` (in which case you don't need to use ```std::atomic```).